opm/material/densead/FastSmallVector.hpp

```diff
@@ -0,0 +1,164 @@
+#include <array>
+#include <cstring>
+
+namespace Opm {
+namespace DenseAd {
+
+template <typename ValueType, unsigned N>
+class FastSmallVector
+{
+public:
+
+    FastSmallVector();
+    explicit FastSmallVector(const size_t num_elem);
+    FastSmallVector(const size_t num_elem, const ValueType value);
+    FastSmallVector(const FastSmallVector& other);
+    FastSmallVector(FastSmallVector&& other);
+    ~FastSmallVector();
+
+    FastSmallVector& operator=(const FastSmallVector& other);
+    FastSmallVector& operator=(FastSmallVector&& other);
+
+    ValueType& operator[](size_t index);
+    const ValueType& operator[](size_t index) const;
+    size_t size() const;
+
+private:
+    std::array<ValueType, N> small_buffer_;
+    std::size_t size_;
+    ValueType* data_ptr_;
+};
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector()
+    : size_(0)
+    , data_ptr_(small_buffer_.data())
+{
+}
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector(const size_t num_elem)
+    : size_(num_elem)
+    , data_ptr_(small_buffer_.data())
+{
+    if (size_ > N) {
+        data_ptr_ = new ValueType[num_elem];
```

Comment:
in `opm-material`, use camelCase instead of_underscores please. (this occurs quite often in the file. if you want, I'll make a patch for you.)

Reply:
I think I did not use camelCase for variables for this file at all unless I oversaw it. Please make a patch as you like.  To me, it looks bad when variables and functions look similar. anyway, it is something of personal taste.

Reply:
> anyway, it is something of personal taste.

sure. so far, the OPM modus-operanti was to use the conventions of the respective module.

Anyway, I'll make a cleanup patch for this but then I expect you to do the same for patches that affect code which you maintain in the future.

Reply:
If you have not began, you can do it after I finished addressing or after this PR. It is just more efficient.

Reply:
too late: https://github.com/GitPaean/opm-material/pull/1

just merge this into your PR and I'm happy with the style of `FastSmallVector`.